### PR TITLE
refactor(elixir): `TypedCBOR.encode()` returns a tuple on success and…

### LINKED ISSUE #3189: TypedCBOR: `encode` wrapping exception does not return `{:ok ...}` tuple
There are two functions defined in `Ockam.TypedCBOR` for encoding: `encode` and `encode!`.

Former is calling `wrap_exception(&encode!...` which will return `{:error, reason}` if there is an exception in `encode!` instead of throwing the exception.

But if there is no exception, the return value is `binary()` and not `{:ok, binary()}` and the function return value is used directly in the code.

This means that if we encounter exception, there will be `{:error, reason}` tuples in the data instead of encoded values because it's not checked anywhere.

As a solution we can switch all current usage of `encode` to `encode!` and fix the function to return `{:ok, binary()}` on success instead of `binary()` so we can use `encode` in the future code.

### PR DESCRIPTION
**Pull request by davidag**
… failure

Fixes #3189

## Current Behavior

`TypedCBOR.encode()` returns a value on success and a tuple `{:error, reason}` on error.

## Proposed Changes

Make `TypedCBOR.encode()` return a tuple `{:ok, data}` on success in symmetry with `TypedCBOR.decode()`. Solution described in #3189

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

### CODE REVIEW

**Comment by hairyhum** [comment]
@davidag thank you for the contribution. Can you also please check usages of `encode` for `Ockam.API.Discovery.ServiceInfo`, `Ockam.API.Request` and `Ockam.API.Response`?

**Comment by davidag** [comment]
@hairyhum Updated so that all three use `encode!()` instead of `encode()`, which I think is better because now errors will raise.

File: implementations/elixir/ockam/ockam/lib/ockam/api/discovery/service_info.ex [hunk 1]
```diff
@@ -48,7 +48,7 @@ defmodule Ockam.API.Discovery.ServiceInfo do
   end
 
   def encode(%__MODULE__{} = service_info) do
```
Comment (hairyhum) [inline implementations/elixir/ockam/ockam/lib/ockam/api/discovery/service_info.ex]: For consistency this should also be called `encode!`

File: implementations/elixir/ockam/ockam/lib/ockam/api/request.ex [hunk 2]
```diff
@@ -38,7 +38,7 @@ defmodule Ockam.API.Request do
       end
 
     base =
-      Header.encode(%Header{
```
Comment (hairyhum) [inline implementations/elixir/ockam/ockam/lib/ockam/api/request.ex:41]: Let's rename `Request.encode/1` -> `Request.encode!/1` for consistency. We can address mixed use of `:bare` and `CBOR` later.

File: implementations/elixir/ockam/ockam/lib/ockam/api/response.ex [hunk 3]
```diff
@@ -36,7 +36,7 @@ defmodule Ockam.API.Response do
       end
 
     base =
-      Header.encode(%Header{
```
Comment (hairyhum) [inline implementations/elixir/ockam/ockam/lib/ockam/api/response.ex:39]: Let's rename `Response.encode/1` -> `Response.encode!/1` for consistency. We can address mixed use of `:bare` and `CBOR` later.

**Review by hairyhum — CHANGES_REQUESTED** [review]

File: implementations/elixir/ockam/ockam/lib/ockam/api/response.ex (hunk 3 above)
Comment (davidag) [reply implementations/elixir/ockam/ockam/lib/ockam/api/response.ex:39]: Good point, I've changed all three 👌🏼

**Comment by hairyhum** [comment]
I've just found that update in return type of `encode/1` changes the return type of `encode_list/1` for the modules using typed cbor plugin.
@davidag  Can you please check and update usages of `encode_list/1` to `encode_list!/1`?

**Comment by davidag** [comment]
@hairyhum fixed one usage of `encode_list()` and rebased. Let me know if there's anything else missing 🙏🏼

**Review by hairyhum — DISMISSED** [review]

**Review by hairyhum — APPROVED** [review]

**Pull request merged**